File: ts/indicators/mark-jurik/jurik-wavelet-sampler/jurik-wavelet-sampler.ts
import { buildMetadata } from '../../core/build-metadata';
import { componentTripleMnemonic } from '../../core/component-triple-mnemonic';
import { IndicatorMetadata } from '../../core/indicator-metadata';
import { IndicatorIdentifier } from '../../core/indicator-identifier';
import { LineIndicator } from '../../core/line-indicator';
import { JurikWaveletSamplerParams } from './params';

/** Function to calculate mnemonic of a JurikWaveletSampler indicator. */
export const jurikWaveletSamplerMnemonic = (params: JurikWaveletSamplerParams): string =>
    'jwav('.concat(
        params.index.toString(),
        componentTripleMnemonic(params.barComponent, params.quoteComponent, params.tradeComponent),
        ')');

/** (n, M) parameters for each wavelet column. */
const nmTable: Array<{ n: number; m: number }> = [
    { n: 1, m: 0 }, { n: 2, m: 0 }, { n: 3, m: 0 }, { n: 4, m: 0 }, { n: 5, m: 0 },
    { n: 7, m: 2 }, { n: 10, m: 2 }, { n: 14, m: 4 }, { n: 19, m: 4 }, { n: 26, m: 8 },
    { n: 35, m: 8 }, { n: 48, m: 16 }, { n: 65, m: 16 }, { n: 90, m: 32 }, { n: 123, m: 32 },
    { n: 172, m: 64 }, { n: 237, m: 64 }, { n: 334, m: 128 },
];

/** Jurik Wavelet Sampler (WAV) line indicator. */
export class JurikWaveletSampler extends LineIndicator {
    private readonly index: number;
    private readonly maxLookback: number;
    private prices: number[] = [];
    private barCount = 0;
    private readonly cols: number[];

    public constructor(params: JurikWaveletSamplerParams) {
        super();

        if (params.index < 1 || params.index > 18) {
            throw new Error('invalid jurik wavelet sampler parameters: index must be in range [1, 18]');
        }

        this.index = params.index;

        // Compute max lookback.
        let maxLookback = 0;
        for (let c = 0; c < params.index; c++) {
            const lb = nmTable[c].n + Math.trunc(nmTable[c].m / 2);
            if (lb > maxLookback) { maxLookback = lb; }
        }
        this.maxLookback = maxLookback;
        this.cols = new Array(params.index).fill(0);
        this.mnemonic = jurikWaveletSamplerMnemonic(params);
        this.description = 'Jurik wavelet sampler ' + this.mnemonic;
        this.primed = false;
        this.barComponent = params.barComponent;
        this.quoteComponent = params.quoteComponent;
        this.tradeComponent = params.tradeComponent;
    }

    /** Describes the output data of the indicator. */
    public metadata(): IndicatorMetadata {
        return buildMetadata(
            IndicatorIdentifier.JurikWaveletSampler,
            this.mnemonic,
            this.description,
            [{ mnemonic: this.mnemonic, description: this.description }],
        );
    }

    /** Returns the current column values after the last update. */
    public columns(): number[] {
        return this.cols.slice();
    }

    /** Updates the value of the WAV indicator given the next sample. */
    public update(sample: number): number {
        if (Number.isNaN(sample)) {
            return sample;
        }

        this.prices.push(sample);
        this.barCount++;

        let allValid = true;

        for (let c = 0; c < this.index; c++) {
            const n = nmTable[c].n;
            const m = nmTable[c].m;
            const deadZone = n + Math.trunc(m / 2);

            if (this.barCount <= deadZone) {
                this.cols[c] = NaN;
                allValid = false;
            } else {
                if (m === 0) {
                    // Simple lag.
                    this.cols[c] = this.prices[this.barCount - 1 - n];
                } else {
                    // Mean of (M+1) prices centered at lag n.
                    const half = Math.trunc(m / 2);
                    const centerIdx = this.barCount - 1 - n;
                    let total = 0.0;

                    for (let k = centerIdx - half; k <= centerIdx + half; k++) {
                        total += this.prices[k];
                    }

                    this.cols[c] = total / (m + 1);
                }
            }
        }

        if (allValid) {
            this.primed = true;
        }

        // Return first column as the framework output.
        return this.cols[0];
    }
}

File: go/indicators/markjurik/jurikwaveletsampler/jurikwaveletsampler.go
package jurikwaveletsampler

import (
	"fmt"
	"math"
	"sync"

	"zpano/entities"
	"zpano/indicators/core"
)

// nmEntry holds the (n, M) parameters for each wavelet column.
type nmEntry struct {
	n int
	m int
}

// nmTable defines the (n, M) parameters for columns 0..17.
var nmTable = []nmEntry{
	{1, 0}, {2, 0}, {3, 0}, {4, 0}, {5, 0},
	{7, 2}, {10, 2}, {14, 4}, {19, 4}, {26, 8},
	{35, 8}, {48, 16}, {65, 16}, {90, 32}, {123, 32},
	{172, 64}, {237, 64}, {334, 128},
}

// JurikWaveletSampler computes the Jurik wavelet sampler.
// Produces `index` output columns per bar, each representing a different
// multi-resolution scale. The framework output is the first column value.
type JurikWaveletSampler struct {
	mu sync.RWMutex
	core.LineIndicator
	primed      bool
	index       int
	maxLookback int
	prices      []float64
	barCount    int
	columns     []float64
}

// NewJurikWaveletSampler returns an instance of the indicator.
func NewJurikWaveletSampler(p *JurikWaveletSamplerParams) (*JurikWaveletSampler, error) {
	return newJurikWaveletSampler(p.Index, p.BarComponent, p.QuoteComponent, p.TradeComponent)
}

func newJurikWaveletSampler(index int,
	bc entities.BarComponent, qc entities.QuoteComponent, tc entities.TradeComponent,
) (*JurikWaveletSampler, error) {
	const (
		invalid = "invalid jurik wavelet sampler parameters"
		fmts    = "%s: %s"
		fmtw    = "%s: %w"
fmtn = "jwav(%d%s)"
	)

	var (
		err       error
		barFunc   entities.BarFunc
		quoteFunc entities.QuoteFunc
		tradeFunc entities.TradeFunc
	)

	if index < 1 || index > 18 {
		return nil, fmt.Errorf(fmts, invalid, "index must be in range [1, 18]")
	}

	if bc == 0 {
		bc = entities.DefaultBarComponent
	}

	if qc == 0 {
		qc = entities.DefaultQuoteComponent
	}

	if tc == 0 {
		tc = entities.DefaultTradeComponent
	}

	if barFunc, err = entities.BarComponentFunc(bc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	if quoteFunc, err = entities.QuoteComponentFunc(qc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	if tradeFunc, err = entities.TradeComponentFunc(tc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	// Compute max lookback.
	maxLookback := 0

	for c := 0; c < index; c++ {
		lb := nmTable[c].n + nmTable[c].m/2
		if lb > maxLookback {
			maxLookback = lb
		}
	}

	mnemonic := fmt.Sprintf(fmtn, index, core.ComponentTripleMnemonic(bc, qc, tc))
	desc := "Jurik wavelet sampler " + mnemonic

	ind := &JurikWaveletSampler{
		index:       index,
		maxLookback: maxLookback,
		columns:     make([]float64, index),
	}

	ind.LineIndicator = core.NewLineIndicator(mnemonic, desc, barFunc, quoteFunc, tradeFunc, ind.Update)

	return ind, nil
}

// IsPrimed indicates whether the indicator is primed.
func (s *JurikWaveletSampler) IsPrimed() bool {
	s.mu.RLock()
	defer s.mu.RUnlock()

	return s.primed
}

// Metadata describes the output data of the indicator.
func (s *JurikWaveletSampler) Metadata() core.Metadata {
	return core.BuildMetadata(
		core.JurikWaveletSampler,
		s.LineIndicator.Mnemonic,
		s.LineIndicator.Description,
		[]core.OutputText{
			{Mnemonic: s.LineIndicator.Mnemonic, Description: s.LineIndicator.Description},
		},
	)
}

// Columns returns the current column values after the last update.
func (s *JurikWaveletSampler) Columns() []float64 {
	s.mu.RLock()
	defer s.mu.RUnlock()

	result := make([]float64, len(s.columns))
	copy(result, s.columns)

	return result
}

// Update updates the indicator given the next sample.
func (s *JurikWaveletSampler) Update(sample float64) float64 {
	if math.IsNaN(sample) {
		return sample
	}

	s.mu.Lock()
	defer s.mu.Unlock()

	s.prices = append(s.prices, sample)
	s.barCount++

	allValid := true

	for c := 0; c < s.index; c++ {
		n := nmTable[c].n
		m := nmTable[c].m
		deadZone := n + m/2

		if s.barCount <= deadZone {
			s.columns[c] = math.NaN()
			allValid = false
		} else {
			if m == 0 {
				// Simple lag.
				s.columns[c] = s.prices[s.barCount-1-n]
			} else {
				// Mean of (M+1) prices centered at lag n.
				half := m / 2
				centerIdx := s.barCount - 1 - n
				total := 0.0

				for k := centerIdx - half; k <= centerIdx+half; k++ {
					total += s.prices[k]
				}

				s.columns[c] = total / float64(m+1)
			}
		}
	}

	if allValid {
		s.primed = true
	}

	// Return first column as the framework output.
	return s.columns[0]
}
